From bdebad8d32d51c3069f55d8c4b3cd7389b1f30bd Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 10 Feb 2025 13:29:25 +0800
Subject: [PATCH 157/236] mtk: wpa_s: fix MLD connection issue when WPS is used

After AP/STA complete the WPS 8-way handshake, the STA triggers a
de-authentication and a re-association. However, the AP might still
be cleaning up the WPS stuff so that it cannot immediately reply the
STA's ML probe request, which leads to the STA chooses to establish
only single link with the AP.

This commit delays the re-association after the WPS 8-way handshake, and
prevents the association retry in disassociation event handling if
current key management is WPS.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 wpa_supplicant/events.c         | 3 ++-
 wpa_supplicant/wps_supplicant.c | 2 +-
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/wpa_supplicant/events.c b/wpa_supplicant/events.c
index f5bcc21af..616eae4c4 100644
--- a/wpa_supplicant/events.c
+++ b/wpa_supplicant/events.c
@@ -4783,7 +4783,8 @@ static void wpa_supplicant_event_disassoc_finish(struct wpa_supplicant *wpa_s,
 			 */
 			fast_reconnect = wpa_s->current_bss;
 			fast_reconnect_ssid = wpa_s->current_ssid;
-		} else if (wpa_s->wpa_state >= WPA_ASSOCIATING) {
+		} else if (wpa_s->wpa_state >= WPA_ASSOCIATING &&
+			   wpa_s->key_mgmt != WPA_KEY_MGMT_WPS) {
 			wpa_supplicant_req_scan(wpa_s, 0, 100000);
 		} else {
 			wpa_dbg(wpa_s, MSG_DEBUG, "Do not request new "
diff --git a/wpa_supplicant/wps_supplicant.c b/wpa_supplicant/wps_supplicant.c
index 74a708c25..32c4f067c 100644
--- a/wpa_supplicant/wps_supplicant.c
+++ b/wpa_supplicant/wps_supplicant.c
@@ -170,7 +170,7 @@ int wpas_wps_eapol_cb(struct wpa_supplicant *wpa_s)
 		 */
 		wpa_printf(MSG_DEBUG, "WPS: Continue association from timeout");
 		wpas_wps_assoc_with_cred_cancel(wpa_s);
-		eloop_register_timeout(0, 10000,
+		eloop_register_timeout(0, 100000,
 				       wpas_wps_assoc_with_cred, wpa_s,
 				       use_fast_assoc ? (void *) 1 :
 				       (void *) 0);
-- 
2.45.2

